### PR TITLE
feat: incremental elaboration of definition headers, bodies, and tactics

### PR DESCRIPTION
**Pull request by Kha**
Extends Lean's incremental reporting and reuse between commands into various steps inside declarations:
* headers and bodies of each (mutual) definition/theorem
* `theorem ... := by` for each contained tactic step, including recursively inside supported combinators currently consisting of
  * `·` (cdot), `case`, `next`
  * `induction`, `cases`
  * macros such as `next` unfolding to the above

Recording 2024-05-10 at 11 07 32

*Incremental reuse* means not recomputing any such steps if they are not affected by a document change. *Incremental reporting* includes the parts seen in the recording above: the progress bar and messages. Other language server features such as hover etc. are *not yet* supported incrementally, i.e. they are shown only when the declaration has been fully processed as before.

### CODE REVIEW

**Comment by Kha** [comment]
#3636 was accidentally closed without merging

File: src/Lean/Elab/MutualDef.lean [hunk 1]
```diff
@@ -68,6 +64,8 @@ private def check (prevHeaders : Array DefViewElabHeader) (newHeader : DefViewEl
     throwError "'partial' theorems are not allowed, 'partial' is a code generation directive"
   if newHeader.kind.isTheorem && newHeader.modifiers.isNoncomputable then
     throwError "'theorem' subsumes 'noncomputable', code is not generated for theorems"
+  if newHeader.modifiers.isNoncomputable && newHeader.modifiers.isUnsafe then
+    throwError "'noncomputable unsafe' is not allowed"
```
Comment (tydeu) [inline src/Lean/Elab/MutualDef.lean:68]: A merge at some point appears to have gone wrong.

Comment (tydeu) [reply src/Lean/Elab/MutualDef.lean:68]: Oh, looking further, the are other changes not incorporate here either. I presume you have just not merged master in a while. Sorry for the noise!

Comment (Kha) [reply src/Lean/Elab/MutualDef.lean]: No, you were completely right! Thanks for flagging.

**Comment by Kha** [comment]
Drafted until the base is changed back to master just before merge

**Pull request merged**